support loader patch for upstream

yetist opened the pull request:
第二波拟提交上游的补丁，支持 efistub 内核启动...

需要配合最新的物理地址内核使用，并且可以看到，此 PR 依赖于上游的 loadfile2 补丁合入主线。

yetist commented on include/grub/loongarch64/linux.h:
```diff
@@ -0,0 +1,42 @@
+#ifndef GRUB_LOONGARCH64_LINUX_HEADER
+#define GRUB_LOONGARCH64_LINUX_HEADER 1
+
+#define GRUB_LINUX_LOONGARCH_MAGIC_SIGNATURE  0x637241676E6F6F4C /* 'LoongArc' */
```
@chenhuacai

目前内核里写的是 “LoongArch"，在 grub 这里就显得特别的别扭。需要内核先确定好，并形成文档，这个补丁才有可能被上游合入。

xen0n replied:
Signature 这种地方，有机会的话我觉得还是玩个梗吧，“龙芯架构”四个字的GBK编码都比LoongArc强

xen0n replied:
我建议是用四字成语的GBK编码（UTF-8的话，8字节只能放两个汉字），比如“神龙摆尾”

xen0n replied:
或者 "Looooong" 这种外国人也能懂的简单玩笑

yetist replied:
已确定为 "Loongson"

xen0n replied:
> 已确定为 "Loongson"

妈的这么无聊的么。。味道淡出天际🌚

xen0n replied:
已经没有这个问题了，magic 整个去掉了

yetist replied:
好，我确认一下

xry111 commented on conf/Makefile.common:
```diff
@@ -17,6 +17,9 @@ endif
   CFLAGS_PLATFORM += -mcmodel=large
 endif
+if COND_loongarch64
+  CPPFLAGS_PLATFORM = -Wa,-mla-global-with-abs -mcmodel=large
```
这里确定不是 CFLAGS_PLATFORM 吗？习惯上 CPPFLAGS 是预处理阶段用的 FLAG，而这个是代码生成阶段的事情。

xry111 commented:
咨询一下，如果要使用新的 grub (这里的或者 dev-la64 里的) 就必须升级固件是吗？

yetist commented:
这个分支只能用物理地址内核

xry111 commented:
> 这个分支只能用物理地址固件

明白了，谢谢！

xry111 commented on grub-core/kern/loongarch64/dl.c:
```diff
@@ -0,0 +1,102 @@
+#include <grub/types.h>
+#include <grub/mm.h>
+#include <grub/i18n.h>
+#include <grub/cpu/reloc.h>
```
should it be `grub/loongarch64/reloc.h`?  Without the change I got a build error here.

yetist replied:
`grub/cpu` 应该会链接到 `grub/<arch>` 目录上，你再看一下？

xry111 replied:
正常构建的时候确实没问题，也是 `make dist` 时候挂了。我来看看为啥……

xry111 replied:
要不问一下上游对 `grub/cpu` 的使用是不是有什么限制？别的架构的 `dl-helper.c` 也直接写了 `grub/<arch>/reloc.h`。

yetist replied:
你是正常make完成之后，再继续make distcheck的吗？感觉不先make，直接make dist可能会这样，如果假设能成立，那x86可能也会这样。

xry111 replied:
应该是只要不在 loongarch 上 `make dist` 就会这样。`cpu` 符号链接是 `configure` 脚本创建的，`make dist` 也不会给它传 `--target` 之类的东西，所以 `make dist` 时候会链接到 `x86` (假设上游用 x86 机器生成压缩包)。

而 `make dist` 会用预处理器检查 libgrubmods_a_SOURCES 和 libgrubkern_a_SOURCES 里面的所有东西，其他架构都保证这两个变量中的所有文件不用 `cpu` 符号链接。但是把 `dl_helper.c` 排除出 libgrubmods_a_SOURCES 好像并不可能……

yetist replied:
也许可以试下不改dl.c，而把dl_helper.c改成include <grub/loongarch64/reloc.h>？

xry111 replied:
坏了，我评论错文件了…… 是要改 dl_helper.c，不是这个 :(

yetist replied:
代码更新了，麻烦再看一下这几个问题，谢谢。

xry111 commented on grub-core/lib/loongarch64/setjmp.S:
```diff
@@ -0,0 +1,69 @@
+/*
```
You need to add
```
extra_dist = lib/loongarch64/setjmp.S;
```

into grub-core/Makefile.core.def (put it below `extra_dist = lib/riscv/setjmp.S;`).  The upstream will generate release tarballs using "make dist", without this line the release tarball will lack setjmp.S and fail to build.

yetist replied:
谢谢，确实 `make distcheck` 会失败。

xen0n commented on configure.ac:
```diff
@@ -854,6 +854,20 @@ if ( test "x$target_cpu" = xi386 || test "x$target_cpu" = xx86_64 ); then
 fi
 
+if test "x$target_cpu" = xloongarch64; then
+  AC_CACHE_CHECK([whether -Wa,-mla-global-with-abs works], [grub_cv_cc_mla_global_with_abs], [
```
会有不支持 `-mla-global-with-abs` 的 LA 汇编器吗？如果不加会如何，怎么总感觉不加会炸啊，除非那个神奇的不支持该选项的汇编器默认行为就是 `-mla-global-with-abs`

yetist replied:
只是重新整理了代码，之前它在这里：https://github.com/loongarch64/grub/pull/6#discussion_r831027490

这里加check的代码，是参考上面x86的，对as做了测试。目前的实际情况是，as支持，但--help不显示。

xen0n commented on grub-core/kern/loongarch64/dl.c:
```diff
@@ -72,19 +73,83 @@ grub_arch_dl_relocate_symbols (grub_dl_t mod, void *ehdr,
+			       "GOT_PC_LO12 without matching GOT_PC_HI2");
+	  unmatched_got_pc_page--;
+	  break;
 	GRUB_LOONGARCH64_RELOCATION (&stack, place, sym_addr)
```
把这个跟上面 sop 的那俩 cases 放一起？现在看代码像是 sop 的处理被一堆 new-style relocs 从中打断了一样。

yetist replied:
感觉没必要。如果放到上面去，通用处理夹杂在特殊处理之间的话，从视觉上看反而有点怪异。

xen0n commented on util/grub-mkimagexx.c:
```diff
@@ -1143,6 +1152,62 @@ SUFFIX (relocate_addrs) (Elf_Ehdr *e, struct section_metadata *smd,
+		     break;
+		   case R_LARCH_GOT_PC_LO12:
+		     if (unmatched_adr_got_page == 0)
+		       grub_util_error ("GOT_PC_LO12 without matching GOT_PC_HI2");
```
`GOT_PC_HI20`

xry111 commented on grub-core/loader/efi/linux.c:
```diff
@@ -48,10 +48,19 @@ static grub_uint32_t cmdline_size;
+  struct grub_pe32_coff_header *coff_header;
+  struct grub_pe32_optional_header *optional_header;
+
+  if (lh->magic != GRUB_LINUX_ARCH_MAGIC_SIGNATURE)
```
内核那边反馈[建议把这个东西鲨掉](https://lore.kernel.org/loongarch/CAAhV-H5CXeG9mNxqJLouvSGLqno4DSwbpPOO5xG2D6ptF2dSTQ@mail.gmail.com/T/#m7b3fdf696c71a8eeac171fe0b9528648e7c5332c)，[新的 EFI stub patch](https://lore.kernel.org/loongarch/20220818030707.2836607-1-chenhuacai@loongson.cn/T/#u) 也是按没有这东西写的。

xen0n commented:
commit c25adce6f8a8ed320ba54cc3ed9fc4d2e0ea4f90 在我这也不行，我基本放弃在 A2101 测试这些了

Commit:
term/at_keyboard: Add timeout to fix hang on HP EliteBooks

This fixes the GRUB on Coreboot on HP EliteBooks by implementing
a 200 ms timeout. The GRUB used to hang.

Fixes: https://ticket.coreboot.org/issues/141

Signed-off-by: Riku Viitanen <riku.viitanen@protonmail.com>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>

frozencemetery committed:
Revert "net/http: Allow use of non-standard TCP/IP ports"

The notation introduced in ac8a37dda (net/http: Allow use of non-standard
TCP/IP ports) contradicts that used in downstream distributions including
Fedora, RHEL, Debian, Ubuntu, and others. Revert it and apply the downstream
notation which was originally proposed to the GRUB in 2016.

This reverts commit ac8a37dda (net/http: Allow use of non-standard TCP/IP ports).

Signed-off-by: Robbie Harwood <rharwood@redhat.com>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>

frozencemetery committed:
net: Read bracketed IPv6 addrs and port numbers

Allow specifying port numbers for http and tftp paths and allow IPv6
addresses to be recognized with brackets around them, which is required
to specify a port number.

Co-authored-by: Aaron Miller <aaronmiller@fb.com>
Signed-off-by: Aaron Miller <aaronmiller@fb.com>
Co-authored-by: Peter Jones <pjones@redhat.com>
Signed-off-by: Peter Jones <pjones@redhat.com>
Signed-off-by: Robbie Harwood <rharwood@redhat.com>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>

yetist commented:
根据上游 review 意见，帮忙检查一下，看看是否有漏掉的修改，谢谢。
https://lists.gnu.org/archive/html/grub-devel/2023-04/msg00142.html

chrisccoulson committed:
font: Try opening fonts from the bundled memdisk

GRUB since 93a786a00 (kern/efi/sb: Enforce verification of font files)
has enforced verification of font files in secure boot mode. In order to
continue to be able to load some default fonts, vendors may bundle them
with their signed EFI image by adding them to the built-in memdisk.

This change makes the font loader try loading fonts from the memdisk
before the prefix path when attempting to load a font file by specifying
its filename, which avoids having to make changes to GRUB configurations
in order to accommodate memdisk bundled fonts. It expects the directory
structure to be the same as fonts stored in the prefix path,
i.e. /fonts/<name>.pf2.

Signed-off-by: Chris Coulson <chris.coulson@canonical.com>
Reviewed-by: Steve McIntyre <93sam@debian.org>
Tested-by: Steve McIntyre <93sam@debian.org>
Reviewed-by: Robbie Harwood <rharwood@redhat.com>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>

yetist committed:
pe: Add LoongArch definitions

Add PE machine types [1] and relocation types [2] for LoongArch to
the current in-repo definitions.

[1] https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#machine-types
[2] https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#base-relocation-types

Signed-off-by: Zhou Yang <zhouyang@loongson.cn>
Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>

yetist committed:
elf: Add LoongArch definitions

Add ELF e_machine ID [1] and relocations types [2] for LoongArch to
the current in-repo definitions.

[1] https://loongson.github.io/LoongArch-Documentation/LoongArch-ELF-ABI-EN.html#_e_machine_identifies_the_machine
[2] https://loongson.github.io/LoongArch-Documentation/LoongArch-ELF-ABI-EN.html#_relocations

Signed-off-by: Zhou Yang <zhouyang@loongson.cn>
Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>

yetist committed:
loongarch: Add setjmp implementation

This patch adds a setjmp implementation for LoongArch.

Signed-off-by: Zhou Yang <zhouyang@loongson.cn>
Signed-off-by: Sun Haiyong <sunhaiyong@loongson.cn>
Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>

yetist committed:
loongarch: Add early startup code

On entry, we need to save the system table pointer as well as our image
handle. Add an early startup file that saves them and then brings us
into our main function.

Signed-off-by: Zhou Yang <zhouyang@loongson.cn>
Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>

yetist committed:
loongarch: Add support for ELF psABI v1.00 relocations

This patch adds support of the stack-based LoongArch relocations
throughout GRUB, including tools, dynamic linkage, and support for
conversion of ELF relocations into PE ones. A stack machine is required
to handle these per the spec [1] (see the R_LARCH_SOP types), of which
a simple implementation is included.

These relocations are produced by binutils 2.38 and 2.39, while the newer
v2.00 relocs require more recent toolchain (binutils 2.40+ & gcc 13+, or
LLVM 16+). GCC 13 has not been officially released as of early 2023, so
support for v1.00 relocs are expected to stay relevant for a while.

[1] https://loongson.github.io/LoongArch-Documentation/LoongArch-ELF-ABI-EN.html#_relocations

Signed-off-by: Zhou Yang <zhouyang@loongson.cn>
Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>

yetist committed:
loongarch: Add support for ELF psABI v2.00 relocations

A new set of relocation types was added in the LoongArch ELF psABI v2.00
spec [1], [2] to replace the stack-based scheme in v1.00. Toolchain
support is available from binutils 2.40 and gcc 13 onwards.

This patch adds support for the new relocation types, that are simpler
to handle (in particular, stack operations are gone). Support for the
v1.00 relocs are kept for now, for compatibility with older toolchains.

[1] https://github.com/loongson/LoongArch-Documentation/pull/57
[2] https://loongson.github.io/LoongArch-Documentation/LoongArch-ELF-ABI-EN.html#_appendix_revision_history

Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>

yetist committed:
loongarch: Add auxiliary files

Add support for manipulating architectural cache and timers, and EFI
memory maps.

Signed-off-by: Zhou Yang <zhouyang@loongson.cn>
Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>

yetist committed:
loongarch: Add to build system

This patch adds LoongArch to the GRUB build system and various tools,
so GRUB can be built on LoongArch as a UEFI application.

Signed-off-by: Zhou Yang <zhouyang@loongson.cn>
Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>

yetist committed:
tests: Fix timezone inconsistency in squashfs_test

The image timestamp was not returned in UTC, but the following logic
expected and used UTC.

This patch fixes the test failure like described below:

  unsquashfs -s /tmp/grub-fs-tester.20230407111703613257436.squash4_gzip.9R4/squash4_gzip_512_4096_1_0.img
  grep '^Creation'
  awk '{print $6 " " $7 " " $8 " " $9 " " $10; }'
  FSTIME='Fri Apr 7 11:17:05 2023'
  date -d 'Fri Apr 7 11:17:05 2023' -u '+%Y-%m-%d %H:%M:%S'
  FSTIME='2023-04-07 11:17:05'
  date -d '2023-04-07 11:17:05 UTC -1 second' -u '+%Y-%m-%d %H:%M:%S'
  FSTIMEM1='2023-04-07 11:17:04'
  date -d '2023-04-07 11:17:05 UTC -2 second' -u '+%Y-%m-%d %H:%M:%S'
  FSTIMEM2='2023-04-07 11:17:03'
  date -d '2023-04-07 11:17:05 UTC -3 second' -u '+%Y-%m-%d %H:%M:%S'
  FSTIMEM3='2023-04-07 11:17:02'
  grep -F 'Last modification time 2023-04-07 11:17:05'
  echo 'Device loop0: Filesystem type squash4 - Last modification time 2023-04-07 03:17:05 Friday - Sector size 512B - Total size 10680KiB'
  echo 'Device loop0: Filesystem type squash4 - Last modification time 2023-04-07 03:17:05 Friday - Sector size 512B - Total size 10680KiB'
  grep -F 'Last modification time 2023-04-07 11:17:04'
  echo 'Device loop0: Filesystem type squash4 - Last modification time 2023-04-07 03:17:05 Friday - Sector size 512B - Total size 10680KiB'
  grep -F 'Last modification time 2023-04-07 11:17:03'
  echo 'Device loop0: Filesystem type squash4 - Last modification time 2023-04-07 03:17:05 Friday - Sector size 512B - Total size 10680KiB'
  grep -F 'Last modification time 2023-04-07 11:17:02'
  echo FSTIME FAIL

Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>
Reviewed-by: Glenn Washburn <development@efficientek.com>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>

yetist committed:
tests: Add LoongArch to various test cases

I ran the test suite on a 3A5000 desktop, a LoongArch architecture machine,
using Archlinux for LoongArch distro, see https://github.com/loongarchlinux.

Some software versions are:
* linux 6.3.0-rc4
* gcc 13.0.1 20230312
* binutils 2.40
* qemu 7.2.0

The test results of running "make check" with qemu 7.2 are as follows:

=================================
   GRUB 2.11: ./test-suite.log
=================================

  # TOTAL: 85
  # PASS:  73
  # SKIP:  8
  # XFAIL: 0
  # FAIL:  2
  # XPASS: 0
  # ERROR: 2

.. contents:: :depth: 2

ERROR: f2fs_test
================

mount: /tmp/grub-fs-tester.20230418175640563815408.f2fs.UDs/f2fs_rw: unknown filesystem type 'f2fs'.
       dmesg(1) may have more information after failed mount system call.
MOUNT FAILED.
ERROR f2fs_test (exit status: 99)

FAIL: hfs_test
==============

recode: Request `utf8..macroman' is erroneous
mkfs.hfs: name required with -v option
FAIL hfs_test (exit status: 1)

ERROR: zfs_test
===============

zpool not installed; cannot test zfs.
ERROR zfs_test (exit status: 99)

SKIP: pata_test
===============

SKIP pata_test (exit status: 77)

SKIP: ahci_test
===============

SKIP ahci_test (exit status: 77)

SKIP: uhci_test
===============

SKIP uhci_test (exit status: 77)

SKIP: ohci_test
===============

SKIP ohci_test (exit status: 77)

SKIP: ehci_test
===============

SKIP ehci_test (exit status: 77)

SKIP: fddboot_test
==================

SKIP fddboot_test (exit status: 77)

SKIP: netboot_test
==================

SKIP netboot_test (exit status: 77)

SKIP: pseries_test
==================

SKIP pseries_test (exit status: 77)

FAIL: grub_func_test
====================

WARNING: Image format was not specified for '/tmp/grub-shell.HeTAD8Ty3U/grub.iso' and probing guessed raw.
         Automatically detecting the format is dangerous for raw images, write operations on block 0 will be restricted.
         Specify the 'raw' format explicitly to remove the restrictions.
Functional test failure: shift_test:
...
gfxterm_menu_640x480xi16:3 failed: 0xce34981e vs 0xd9f04953
 tests/video_checksum.c:checksum:615: assert failed: 0 Checksum
gfxterm_menu_640x480xi16:2 failed: 0xa8fb749d vs 0xbf3fa5d0
 tests/video_checksum.c:checksum:615: assert failed: 0 Checksum
gfxterm_menu_640x480xi16:1 failed: 0xce34981e vs 0xd9f04953
gfxterm_menu: FAIL
...
videotest_checksum:
videotest_checksum: PASS
exfctest:
exfctest: PASS
TEST FAILURE
FAIL grub_func_test (exit status: 1)

We got 2 errors:

* f2fs_test
The kernel uses 16k pages, causing failures when loading the f2fs kernel module,
see https://github.com/torvalds/linux/blob/master/fs/f2fs/super.c#L4670
This error can be ignored.

* zfs_test
zfs does not support the LoongArch architecture and is not compatible with the
6.3 kernel.
This error can be ignored.

We got 2 failures:

* hfs_test
I use recode 3.7.14-1 on Archlinux, running `recode -l` gives no output `MacRoman`,
so we get this error.
On Linux systems that support LoongArch, there is currently no need to use HFS,
so this failure can be ignored.

* grub_func_test
I don't know the reason for this failure. I guess it may be related to qemu's edk2.
In the previous review, I was told that the failure here is the expected behavior.
So, we can ignore this failure.

Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>

cnmushiba commented on grub-core/kern/loongarch64/dl_helper.c:
```diff
@@ -0,0 +1,260 @@
+}
+
+static void
+grub_loongarch64_stack_push (grub_loongarch64_stack_t stack, grub_uint64_t x)
```
https://www.gnu.org/software/grub/manual/grub-dev/grub-dev.html#Coding-style
2.2 If a function is local, its name may not start with any prefix. It must start with a verb.
不过问题也不大

xen0n commented:
Merged, congrats!

xen0n commented:
Mailing list archive: https://lists.gnu.org/archive/html/grub-devel/2023-05/msg00089.html
